### PR TITLE
feat(rpc): implement getSlot RPC method

### PR DESCRIPTION
**Pull request by ajw221**
### Intent
- Implement getSlot RPC method
### Implementation
- Added ForkChoiceProcessedSlot and OptimisticallyConfirmedSlot atomic trackers to src/replay/trackers.zig
- Added SlotHookContext to src/rpc/methods.zig which reads slot state from SlotTracker based on commitment level
- Updated SlotTracker.root to be atomic for thread-safe access from RPC
- Wired latest_processed_slot updates in consensus mode (during vote handling) and bypass mode (from SlotTree.tip())
- Wired latest_confirmed_slot updates in vote listener when optimistic confirmation threshold is reached
### Ramifications
- ForkChoiceProcessedSlot uses store() rather than fetchMax() because the processed slot can decrease when fork choice switches to a different fork
- In bypass-consensus mode, processed slot is set to the highest fork tip, which differs from Agave's semantics where it only updates when voting
### Tests
- Updated existing unit tests to include the new slot tracker dependencies
- Added assertions verifying slot trackers remain at initial state after expected failures

### CODE REVIEW

File: src/rpc/methods.zig [hunk 1]
```diff
@@ -686,3 +686,58 @@ pub const common = struct {
...
+        return slot;
+    }
+
+    fn getLatestFinalizedSlot(self: @This()) !Slot {
+        const slot = self.slot_tracker.root.load(.monotonic);
+        if (slot == 0) {
+            return error.RpcNoFinalizedSlot;
+        }
+        return slot;
+    }
```
Comment (prestonsn) [inline src/rpc/methods.zig]: check if we need these broken up so redundantly like this

Comment (prestonsn) [reply src/rpc/methods.zig]: Could be useful for other methods, but I'd prefer to pass in a commitment enum arg.

Comment (prestonsn) [reply src/rpc/methods.zig]: Fixed in https://github.com/Syndica/sig/commit/40fae043fa0a03a0d9101fecd414913ea92c67ed

File: src/replay/consensus/process_result.zig [hunk 2]
```diff
@@ -305,9 +305,15 @@ const TestReplayStateResources = struct {
...
+        var latest_confirmed_slot: sig.replay.trackers.OptimisticallyConfirmedSlot = .{};
+        const root: std.atomic.Value(Slot) = .init(0);
+
         self.slot_tracker = SlotTracker{
             .slots = .empty,
-            .root = 0,
+            .latest_processed_slot = &latest_processed_slot,
+            .latest_confirmed_slot = &latest_confirmed_slot,
+            .root = root,
         };
```
Comment (dnut) [inline src/replay/consensus/process_result.zig]: the stack pointers will be invalid after this function returns

File: src/replay/service.zig [hunk 3]
```diff
@@ -1254,6 +1282,8 @@ pub const DependencyStubs = struct {
 
             .replay_threads = 1,
             .stop_at_slot = null,
+            .latest_processed_slot = &latest_processed_slot,
+            .latest_confirmed_slot = &latest_confirmed_slot,
```
Comment (dnut) [inline src/replay/service.zig]: returning stack pointers

File: src/replay/service.zig [hunk 4]
```diff
@@ -1327,6 +1360,8 @@ pub const DependencyStubs = struct {
 
             .replay_threads = num_threads,
             .stop_at_slot = null,
+            .latest_processed_slot = &latest_processed_slot,
+            .latest_confirmed_slot = &latest_confirmed_slot,
```
Comment (dnut) [inline src/replay/service.zig]: returning stack pointers

File: src/consensus/vote_listener.zig [hunk 5]
```diff
@@ -1995,6 +2027,11 @@ test "check trackers" {
         expected_trackers.items,
         actual_trackers.items,
     );
+
+    // Votes were processed but no stake was delegated to validators, so
+    // optimisitic confirmation was not reached.
+    try std.testing.expectEqual(0, confirmed_slot.get());
+    try std.testing.expectEqual(0, confirmed_slot.get());
```
Comment (dnut) [inline src/consensus/vote_listener.zig]:
```suggestion
    try std.testing.expectEqual(0, confirmed_slot.get());
```

File: src/replay/consensus/core.zig [hunk 6]
```diff
@@ -4930,12 +4954,11 @@ test "edge cases - duplicate confirmed slot" {
...
     const progress_map = &replay_state.progress_map;
-    std.debug.assert(slot_tracker.root == 0);
+    const root_slot0 = slot_tracker.root.load(.monotonic);
+    std.debug.assert(root_slot0 == 0);
 
-    const root_slot0 = slot_tracker.root;
     const root_slot0_hash = slot_tracker.getRoot().state.hash.readCopy().?;
-
-    std.debug.assert(root_slot0 == 0); // assert initial root value
+    std.debug.assert(root_slot0_hash.eql(.ZEROES)); // assert initial root hash
```
Comment (dnut) [inline src/replay/consensus/core.zig:4938]: redundant with line below

File: src/consensus/replay_tower.zig [hunk 7]
```diff
@@ -5453,10 +5453,16 @@ pub const TestFixture = struct {
...
-                .constants = constants,
-                .state = state,
-            });
+            var latest_processed_slot: sig.replay.trackers.ForkChoiceProcessedSlot = .{};
+            var latest_confirmed_slot: sig.replay.trackers.OptimisticallyConfirmedSlot = .{};
+
+            break :blk try .init(
+                allocator,
+                &latest_processed_slot,
+                &latest_confirmed_slot,
```
Comment (dnut) [inline src/consensus/replay_tower.zig]: returning stack pointers

File: src/rpc/methods.zig [hunk 8]
```diff
@@ -686,3 +686,26 @@ pub const common = struct {
...
+        const slot = self.slot_tracker.getSlotForCommitment(commitment);
+        if (config.minContextSlot) |min_slot| {
+            if (slot < min_slot) {
+                return error.RpcMinContextSlotNotMet;
+            }
+        }
+        return slot;
+    }
+
+    pub fn getSlot(self: @This(), _: std.mem.Allocator, params: GetSlot) !GetSlot.Response {
```
Comment (dnut) [inline src/rpc/methods.zig]: in general you should use the struct name if it exists, not `@This()`. if the name does not exist (e.g. generics), you can define an alias
```suggestion
    pub fn getSlot(self: SlotHookContext, _: std.mem.Allocator, params: GetSlot) !GetSlot.Response {
```

File: src/rpc/methods.zig [hunk 9]
```diff
@@ -686,3 +686,26 @@ pub const common = struct {
         shredVersion: ?u16 = null,
     };
 };
+
+pub const SlotHookContext = struct {
+    slot_tracker: *const sig.replay.trackers.SlotTracker,
+
+    fn getSlotImpl(
```
Comment (dnut) [inline src/rpc/methods.zig]: this function feels redundant. would you mind inlining it into getSlot?

File: src/replay/trackers.zig [hunk 10]
```diff
@@ -10,6 +10,34 @@ const EpochSchedule = sig.core.EpochSchedule;
...
+        return self.slot.load(.monotonic);
+    }
+};
+
+pub const OptimisticallyConfirmedSlot = struct {
+    slot: std.atomic.Value(Slot) = .init(0),
+
+    pub fn update(self: *@This(), new_slot: Slot) void {
+        _ = self.slot.fetchMax(new_slot, .monotonic);
+    }
```
Comment (dnut) [inline src/replay/trackers.zig]: In theory, this number can go down. It's extremely unlikely to go down, but there's no guarantee that it increases monotonically.

File: src/cmd.zig [hunk 11]
```diff
@@ -1460,6 +1460,9 @@ fn validator(
     else
         null;
 
+    var latest_processed_slot: sig.replay.trackers.ForkChoiceProcessedSlot = .{};
+    var latest_confirmed_slot: sig.replay.trackers.OptimisticallyConfirmedSlot = .{};
```
Comment (dnut) [inline src/cmd.zig]: I don't think these need to be allocated all the way up here. I would push the ownership into one of the structs that is currently holding a pointer to it. I think it makes the most sense for SlotTracker to own the data if it can be done cleanly.

File: src/replay/trackers.zig [hunk 12]
```diff
@@ -10,6 +10,34 @@ const EpochSchedule = sig.core.EpochSchedule;
...
 const SlotState = sig.core.SlotState;
+const Commitment = sig.rpc.methods.common.Commitment;
+
+pub const ForkChoiceProcessedSlot = struct {
+    slot: std.atomic.Value(Slot) = .init(0),
+
+    /// Set the current processed slot (heaviest fork tip).
+    /// Uses store() because this can decrease when the fork choice
+    /// switches to a different fork with a lower slot.
+    pub fn set(self: *@This(), new_slot: Slot) void {
```
Comment (dnut) [inline src/replay/trackers.zig]:
```suggestion
    pub fn set(self: *ForkChoiceProcessedSlot, new_slot: Slot) void {
```

similar pattern applies throughout the pr

https://github.com/Syndica/sig/blob/main/docs/docusaurus/docs/contributing/style-guide.mdx#method-parameters

**Review by dnut — CHANGES_REQUESTED** [review]
my main overall thought is to move all the slot integers to be owned by the slot tracker which would solve several of my comments including the stack pointer uafs

File: src/replay/trackers.zig (hunk 12 above)
Comment (prestonsn) [reply src/replay/trackers.zig]: i should have started here, thanks :)

File: src/consensus/vote_listener.zig (hunk 5 above)
Comment (prestonsn) [reply src/consensus/vote_listener.zig]: Oops, first one should be `processed`

File: src/replay/consensus/process_result.zig (hunk 2 above)
Comment (prestonsn) [reply src/replay/consensus/process_result.zig]: https://github.com/Syndica/sig/pull/1197/commits/ccb5d8b65be0611196677f834c76f56039c81f57

File: src/replay/service.zig (hunk 3 above)
Comment (prestonsn) [reply src/replay/service.zig]: https://github.com/Syndica/sig/pull/1197/commits/ccb5d8b65be0611196677f834c76f56039c81f57

File: src/replay/service.zig (hunk 4 above)
Comment (prestonsn) [reply src/replay/service.zig]: https://github.com/Syndica/sig/pull/1197/commits/ccb5d8b65be0611196677f834c76f56039c81f57

File: src/consensus/vote_listener.zig (hunk 5 above)
Comment (prestonsn) [reply src/consensus/vote_listener.zig]: https://github.com/Syndica/sig/pull/1197/commits/ccb5d8b65be0611196677f834c76f56039c81f57

File: src/consensus/replay_tower.zig (hunk 7 above)
Comment (prestonsn) [reply src/consensus/replay_tower.zig]: https://github.com/Syndica/sig/commit/ccb5d8b65be0611196677f834c76f56039c81f57

File: src/replay/consensus/core.zig (hunk 6 above)
Comment (prestonsn) [reply src/replay/consensus/core.zig:4938]: https://github.com/Syndica/sig/pull/1197/commits/e95e1dbfc63837c5bc8eea8c2d566ac054de707c

File: src/rpc/methods.zig (hunk 9 above)
Comment (prestonsn) [reply src/rpc/methods.zig]: https://github.com/Syndica/sig/pull/1197/commits/e792f674c7847fe8fe09fb84e0c5ce4e7ecd3b3c

File: src/cmd.zig (hunk 11 above)
Comment (prestonsn) [reply src/cmd.zig]: Agreed, this also worked out cleaner: https://github.com/Syndica/sig/pull/1197/commits/ccb5d8b65be0611196677f834c76f56039c81f57

File: src/replay/trackers.zig (hunk 12 above)
Comment (prestonsn) [reply src/replay/trackers.zig]: https://github.com/Syndica/sig/pull/1197/commits/2135f2ffc805cd898b3cd1f114af3e681fd75899

File: src/rpc/methods.zig (hunk 8 above)
Comment (prestonsn) [reply src/rpc/methods.zig]: https://github.com/Syndica/sig/pull/1197/commits/2135f2ffc805cd898b3cd1f114af3e681fd75899

File: src/replay/trackers.zig (hunk 10 above)
Comment (prestonsn) [reply src/replay/trackers.zig]: I was mostly looking at this from the perspective of how Agave handles updating their optimistically confirmed bank. All updates of this bank are guarded with `if new_slot > optimistically_confirmed_bank.bank.slot()`: 

* https://github.com/anza-xyz/agave/blob/v3.1.8/rpc/src/optimistically_confirmed_bank_tracker.rs#L299-L301
* https://github.com/anza-xyz/agave/blob/v3.1.8/rpc/src/optimistically_confirmed_bank_tracker.rs#L376-L377
* https://github.com/anza-xyz/agave/blob/v3.1.8/rpc/src/optimistically_confirmed_bank_tracker.rs#L386-L387

So i'm not seeing a codepath in the Agave code where it can decrease (at least not from the perspective of how they are tracking their optimistically confirmed bank). 

I can see a scenario where the current confirmed slot could abruptly decrease after a higher slot already crossed confirmation threshold (network lag on some high stake votes for a lower slot perhaps?)

File: src/rpc/methods.zig [hunk 13]
```diff
@@ -686,3 +686,19 @@ pub const common = struct {
...
+    pub fn getSlot(self: SlotHookContext, _: std.mem.Allocator, params: GetSlot) !GetSlot.Response {
+        const config = params.config orelse common.CommitmentSlotConfig{};
+        const commitment = config.commitment orelse .finalized;
+        const slot = self.slot_tracker.getSlotForCommitment(commitment);
+        if (config.minContextSlot) |min_slot| {
+            if (slot < min_slot) {
+                return error.RpcMinContextSlotNotMet;
+            }
+        }
+        return slot;
```
Comment (Sobeston) [inline src/rpc/methods.zig]: could early return here, e.g. 
```zig
const min_slot = config.minContextSlot orelse return slot;
return if (slot >= min_slot) slot else error.RpcMinContextSlotNotMet;
```

Comment (prestonsn) [reply src/rpc/methods.zig]: clean! https://github.com/Syndica/sig/pull/1197/commits/cf75dc5b2f7ac0c07934e1a74bb63328ca0d6f95

File: src/replay/consensus/core.zig [hunk 14]
```diff
@@ -136,7 +136,7 @@ pub const TowerConsensus = struct {
             signing: sig.identity.SigningKeys,
             account_reader: AccountReader,
             ledger: *sig.ledger.Ledger,
-            slot_tracker: *const SlotTracker,
+            slot_tracker: *SlotTracker,
```
Comment (dnut) [inline src/replay/consensus/core.zig]: this doesn't need to be mutable

File: src/replay/trackers.zig (hunk 10 above)
Comment (dnut) [reply src/replay/trackers.zig]: this means that agave is incorrect but it's not trivial to handle this correctly, and realistically it would probably never actually need to go down unless something very crazy happened to the cluster. it's fine to just target conformance with agave for now

File: src/replay/consensus/core.zig (hunk 14 above)
Comment (prestonsn) [reply src/replay/consensus/core.zig]: https://github.com/Syndica/sig/pull/1197/commits/7d843701a40202a801fab5e619ae2a87ba6bbca4
